Latest Instagram Posts: Support for Multiple User Accounts

Copons opened the pull request:
![https://i.imgur.com/I6GoNjB.gif](https://i.imgur.com/I6GoNjB.gif)

Fixes #15848

#### Changes proposed in this Pull Request:

* Add support for multiple user accounts in the Latest Instagram Posts block.
  * Users with active Instagram tokens, can reuse them straight away.
  * Users without, will need to connect to Instagram through the usual popup.
* Disconnecting the block won't delete the token anymore, but simply remove it from the block attributes.
  * To permanently delete the token, users will need to head to Marketing/Connections in Calypso, or do it directly from Instagram.
* Remove 2 unnecessary endpoints (`GET /access-token` and `DELETE /delete-access-token`), and corresponding helper code.

#### Screenshots

cc @davemart-in and @olesyabrk for suggestions about the placeholder copy, especially the account selector, and the disconnection button in the block sidebar.

User without connected Instagram accounts:
<img width="629" alt="Screenshot 2020-05-19 at 19 28 26" src="https://user-images.githubusercontent.com/2070010/82365720-efb12000-9a08-11ea-8f56-d549202bc650.png">

User with connected Instagram accounts:
<img width="626" alt="Screenshot 2020-05-20 at 11 48 30" src="https://user-images.githubusercontent.com/2070010/82438004-457cdb00-9a90-11ea-8ef6-2357ea7a6360.png">

Disconnection button:
<img width="275" alt="Screenshot 2020-05-20 at 11 48 23" src="https://user-images.githubusercontent.com/2070010/82438308-bc19d880-9a90-11ea-9199-a2f398c140ee.png">

#### Testing instructions:

WPCOM prerequisites: apply D43668-code and D43669-code (hopefully it patches cleanly).

- Smoke test the Latest Instagram Posts block, and make sure the UX feels good.
- Try connecting more than one Instagram account.
- Make sure connecting new accounts is reflected in the account selector.
- Make sure that after disconnecting a block, the previous account is automatically selected in the account selector.
- Make sure blocks connected to different accounts are visible on the front end to visitors.

#### Proposed changelog entry for your changes:
<!-- Please do not leave this empty. If no changelog entry needed, state as such. -->
* N/A

olesyabrk commented:
@Copons here are some suggestions on copy. 

![image](https://user-images.githubusercontent.com/51215741/82378018-5880a400-99d9-11ea-9f0f-b69f86ae9351.png)

- Remove the widget description, it's self-explanatory and is duplicated in the sidebar. 
- Let's align the content to the header, not to the icon 

![image](https://user-images.githubusercontent.com/51215741/82378259-b0b7a600-99d9-11ea-835c-64a4612c005f.png)

Copons commented:
Thanks @olesyabrk for the suggestions!
I've updated some things, but I've decided to leave others as they are for consistency.

---

Placeholder when user already has some connections:
- Removed unnecessary instructions.
- Set the font weight of the radios to normal.

<img width="626" alt="Screenshot 2020-05-20 at 11 48 30" src="https://user-images.githubusercontent.com/2070010/82438004-457cdb00-9a90-11ea-8ef6-2357ea7a6360.png">

Although, when the user doesn't have any connections, I've chosen to leave the "connect to start sharing" instructions, in order to avoid having a "stray" button.
Also, I think we should leave the placeholder content aligned to the left, as it's what Gutenberg Core does (try with an Image block for example).

---

Sidebar account settings:
- Aligned the account name to the left.
- Updated the disconnection copy.

<img width="275" alt="Screenshot 2020-05-20 at 11 48 23" src="https://user-images.githubusercontent.com/2070010/82438308-bc19d880-9a90-11ea-9199-a2f398c140ee.png">

What I've not touched:
- The account name is an external link (opening in a new tab, outside the current site), so it needs the external icon.
- The disconnection doesn't navigate away, but simply "resets" the block to the placeholder. The arrow icon seems a bit out of place to me. 🤔

jeherve commented on extensions/blocks/instagram-gallery/edit.js:
```diff
@@ -183,32 +239,17 @@ const InstagramGalleryEdit = props => {
 					<PanelBody title={ __( 'Account Settings', 'jetpack' ) }>
-						<PanelRow>
-							<span>{ __( 'Account', 'jetpack' ) }</span>
+						<div>
+							{ __( 'Account:', 'jetpack' ) }{ ' ' }
 							<ExternalLink href={ `https://www.instagram.com/${ instagramUser }/` }>
 								@{ instagramUser }
 							</ExternalLink>
```
I am not sure this is going to work on RTL languages, maybe this should be one translatable string with a placeholder.

Copons replied:
Ooh, makes sense, thanks!

Copons replied:
No, wait, now I remembered why I did it in two columns with `PanelRow` before: Guten `i18n` functions don't support React components.

I'll revert this change back to the two columns version. 🤦

scruffian commented:
<img width="677" alt="Screenshot 2020-05-21 at 12 01 53" src="https://user-images.githubusercontent.com/275961/82553048-3747c100-9b5b-11ea-9b67-d7f4cf52d88a.png">

I wonder if we could change the button text here to explain what is happening more clearly... Maybe "Loading your connections…"?

scruffian commented:
The highlighted state of the radio button looks weird:
<img width="373" alt="Screenshot 2020-05-21 at 12 06 33" src="https://user-images.githubusercontent.com/275961/82553214-85f55b00-9b5b-11ea-91e4-2080ce520afa.png">

Also, when I click "Add a new account" it connects to the account I am currently logged in with, which is the one that's already connected. I know this is a limitation we have but I worry that a lot of users will struggle to understand this. Not sure how we can get around it without adding some messaging though....

scruffian commented:
After I disconnect the block, the connection button is disabled for a moment. I'm not sure why this is necessary:
![May-21-2020 12-17-42](https://user-images.githubusercontent.com/275961/82554119-3adc4780-9b5d-11ea-8f5e-b9118ebb03e6.gif)

Copons commented:
> The highlighted state of the radio button looks weird:
> <img alt="Screenshot 2020-05-21 at 12 06 33" width="373" src="https://user-images.githubusercontent.com/275961/82553214-85f55b00-9b5b-11ea-91e4-2080ce520afa.png">

🤔 Weird, I haven't touched its style at all (just removed the bold, but I don't think it's related).

> 
> Also, when I click "Add a new account" it connects to the account I am currently logged in with, which is the one that's already connected. I know this is a limitation we have but I worry that a lot of users will struggle to understand this. Not sure how we can get around it without adding some messaging though....

Yeah, totally agree with this.
Considering there's no way to change IG account from the IG oAuth screen, we are pretty much forced to add a message about it.
Worst thing is, that it's sort of a complicated one to convey in simple words.
What about showing this line under the "add account" option, only when it's selected?

"If you're currently logged in to Instagram on this device, you might need to log out of it first."

Maybe we could also add a "Learn more" link pointing to the block's support page.

> After I disconnect the block, the connection button is disabled for a moment. I'm not sure why this is necessary:
> ![May-21-2020 12-17-42](https://user-images.githubusercontent.com/275961/82554119-3adc4780-9b5d-11ea-8f5e-b9118ebb03e6.gif)

It's loading/refreshing the connections, in the off chance they've changed meanwhile.
It'll become more clear when I'll update the button with the "loading" label you suggested earlier.

scruffian commented on extensions/blocks/instagram-gallery/editor.scss:
```diff
@@ -1,5 +1,18 @@
+	.components-radio-control {
+		margin-bottom: $default-block-margin;
+		label {
+			font-weight: normal;
```
I tried removing this and it looks the same. Is it for a different version of Gutenberg?

Copons replied:
Interesting!
If I remove that line, it looks like this to me:

<img width="613" alt="Screenshot 2020-05-21 at 17 49 03" src="https://user-images.githubusercontent.com/2070010/82583777-6d9d3480-9b8b-11ea-8fb7-7f502f01308f.png">

I've tested on Guten 8.1 and 8.0, and with Guten deactivated, and in all cases I see this style in the dev tools:
```css
.editor-styles-wrapper label {
  font-weight: 600;
}
```

(Actually, I'm not entirely sure I tested on 8.0. I just updated Guten this morning after a while, and it's 8.1 now. Maybe it was a 8.1 beta before.)

scruffian replied:
Maybe it's a theme thing? Doesn't hurt to leave it in :)

Copons commented:
@jeherve apologies for the push, but I discovered that some themes (2020 in this case) set the `p` font family with more specificity than the placeholder fieldset.
I addressed this in 6a91a1f.

| Before | After |
| - | - |
| <img width="630" alt="Screenshot 2020-05-22 at 12 24 03" src="https://user-images.githubusercontent.com/2070010/82663551-cae6c380-9c27-11ea-8cc0-1bd95f95956e.png"> | <img width="628" alt="Screenshot 2020-05-22 at 12 24 41" src="https://user-images.githubusercontent.com/2070010/82663563-ce7a4a80-9c27-11ea-87f6-2555f188df02.png"> |

Since this dismissed the reviews, I'll also take the chance of merging #15887 first and rebase this.

Copons commented:
r207896-wpcom
